src/tools/rbd_mirror/ImageSyncThrottler.h

```diff
@@ -33,56 +31,49 @@ namespace librbd { namespace journal { struct MirrorPeerClientMeta; } }
 namespace rbd {
 namespace mirror {
 
-template <typename> class ImageSync;
+template <typename> class InstanceWatcher;
 
 class ProgressContext;
 
 /**
  * Manage concurrent image-syncs
  */
 template <typename ImageCtxT = librbd::ImageCtx>
-class ImageSyncThrottler : public md_config_obs_t {
+class ImageSyncThrottler {
 public:
 
   typedef librbd::journal::TypeTraits<ImageCtxT> TypeTraits;
   typedef typename TypeTraits::Journaler Journaler;
   typedef librbd::journal::MirrorPeerClientMeta MirrorPeerClientMeta;
 
   ImageSyncThrottler();
-  ~ImageSyncThrottler() override;
+  ~ImageSyncThrottler();
   ImageSyncThrottler(const ImageSyncThrottler&) = delete;
   ImageSyncThrottler& operator=(const ImageSyncThrottler&) = delete;
 
+  void init(InstanceWatcher<ImageCtxT> *instance_watcher);
+
   void start_sync(ImageCtxT *local_image_ctx,
                   ImageCtxT *remote_image_ctx, SafeTimer *timer,
                   Mutex *timer_lock, const std::string &mirror_uuid,
                   Journaler *journaler, MirrorPeerClientMeta *client_meta,
                   ContextWQ *work_queue, Context *on_finish,
                   ProgressContext *progress_ctx = nullptr);
 
-  void cancel_sync(librados::IoCtx &local_io_ctx,
-                   const std::string local_image_id);
-
-  void set_max_concurrent_syncs(uint32_t max);
+  void cancel_sync(const std::string &local_image_id);
 
   void print_status(Formatter *f, std::stringstream *ss);
 
 private:
-  typedef std::pair<int64_t, std::string> PoolImageId;
-
   struct C_SyncHolder;
 
-  void handle_sync_finished(C_SyncHolder *sync_holder);
+  void handle_sync_started(int r, C_SyncHolder *sync_holder);
+  void handle_sync_finished(int r, C_SyncHolder *sync_holder);
 
-  const char **get_tracked_conf_keys() const override;
-  void handle_conf_change(const struct md_config_t *conf,
-                          const std::set<std::string> &changed) override;
-
-  uint32_t m_max_concurrent_syncs;
+  InstanceWatcher<ImageCtxT> *m_instance_watcher = nullptr;
   Mutex m_lock;
-  std::list<C_SyncHolder *> m_sync_queue;
-  std::map<PoolImageId, C_SyncHolder *> m_inflight_syncs;
-
+  std::map<std::string, C_SyncHolder *> m_waiting_syncs;
```

Comment:
Nit: seems like you can remove this collection since once you schedule the sync, you hand the management over to ```InstanceWatcher``` for starting / canceling. This collection just seems to be used for an assert.